File: src/livekit_agent_simulator/script/verify.py
"""Log-based script verify (hard pass/fail without LLM judge)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from .models import ScriptStep, ScriptVerifySpec

def evaluate_script_log(
    events: list[dict],
    steps: list[ScriptStep],
    verify: ScriptVerifySpec | None = None,
    *,
    scenario: Any | None = None,
    project_root: Path | str | None = None,
) -> dict[str, object]:
    """Log-based PASS/FAIL for scripted adaptive scenarios (no LLM judge required)."""
    cues = [
        e
        for e in events
        if e.get("kind") in ("sim.script.cue", "sim.script.wait", "sim.script.hang_up")
    ]
    agent_finals = [e for e in events if e.get("kind") == "transcript.agent.final"]
    user_finals = [e for e in events if e.get("kind") == "transcript.user.final"]
    interruptions = [e for e in events if e.get("kind") == "interruption"]

    checks: list[dict[str, object]] = []

    for step in steps:
        matching = [c for c in cues if c.get("spec", {}).get("step_id") == step.id]
        if not matching:
            checks.append({"step_id": step.id, "pass": False, "reason": "script step not fired"})
            continue
        cue = matching[0]
        spec = cue.get("spec") or {}
        during = bool(spec.get("during_agent_speech"))
        if step.trigger == "agent_speaking" and step.action == "speak" and not during:
            checks.append(
                {
                    "step_id": step.id,
                    "pass": False,
                    "reason": "cue fired but agent was not active speaker",
                }
            )
            continue
        checks.append(
            {
                "step_id": step.id,
                "pass": True,
                "during_agent_speech": during,
                "trigger": step.trigger,
                "action": step.action,
            }
        )

    cue_ms = cues[0]["ts_mono_ms"] if cues else None
    silence_cues = [
        e
        for e in cues
        if (e.get("spec") or {}).get("trigger") == "silence"
        or (e.get("spec") or {}).get("action") == "wait"
    ]
    silence_ms = silence_cues[0]["ts_mono_ms"] if silence_cues else None
    from .models import counts_for_recovery_barge

    barge_cues = []
    for e in cues:
        if e.get("kind") != "sim.script.cue":
            continue
        spec = e.get("spec") or {}
        cls = spec.get("class") or spec.get("interrupt_class")
        cls_s = str(cls) if cls else None
        if counts_for_recovery_barge(
            barge_in=bool(spec.get("barge_in")), interrupt_class=cls_s
        ):
            barge_cues.append(e)
            continue
        # Legacy heuristic: short during-agent cue without class (treat as correction)
        if (
            not cls_s
            and spec.get("during_agent_speech")
            and spec.get("trigger") == "agent_speaking"
            and int(spec.get("waited_ms") or 9999) < 800
        ):
            barge_cues.append(e)
    barge_ms = barge_cues[0]["ts_mono_ms"] if barge_cues else None

    agent_after_cue = (
        sum(1 for e in agent_finals if cue_ms is not None and e.get("ts_mono_ms", 0) >= cue_ms)
        if cue_ms is not None
        else 0
    )
    user_after_cue = (
        sum(1 for e in user_finals if cue_ms is not None and e.get("ts_mono_ms", 0) >= cue_ms)
        if cue_ms is not None
        else 0
    )
    agent_after_silence = (
        sum(
            1
            for e in agent_finals
            if silence_ms is not None and e.get("ts_mono_ms", 0) >= silence_ms
        )
        if silence_ms is not None
        else 0
    )
    agent_after_barge = (
        sum(
            1
            for e in agent_finals
            if barge_ms is not None and e.get("ts_mono_ms", 0) > barge_ms
        )
        if barge_ms is not None
        else 0
    )

    verify = verify or ScriptVerifySpec()
    if verify.min_agent_finals_after_first_cue > 0:
        ok = agent_after_cue >= verify.min_agent_finals_after_first_cue
        checks.append(
            {
                "check": "min_agent_finals_after_first_cue",
                "pass": ok,
                "expected": verify.min_agent_finals_after_first_cue,
                "actual": agent_after_cue,
            }
        )
    if verify.min_user_finals_after_first_cue > 0:
        ok = user_after_cue >= verify.min_user_finals_after_first_cue
        checks.append(
            {
                "check": "min_user_finals_after_first_cue",
                "pass": ok,
                "expected": verify.min_user_finals_after_first_cue,
                "actual": user_after_cue,
            }
        )
    if verify.min_agent_finals_after_silence > 0:
        ok = agent_after_silence >= verify.min_agent_finals_after_silence
        checks.append(
            {
                "check": "min_agent_finals_after_silence",
                "pass": ok,
                "expected": verify.min_agent_finals_after_silence,
                "actual": agent_after_silence,
            }
        )
    if verify.min_agent_finals_after_barge_in > 0:
        ok = agent_after_barge >= verify.min_agent_finals_after_barge_in
        checks.append(
            {
                "check": "min_agent_finals_after_barge_in",
                "pass": ok,
                "expected": verify.min_agent_finals_after_barge_in,
                "actual": agent_after_barge,
            }
        )
    if verify.min_interruptions is not None:
        ok = len(interruptions) >= verify.min_interruptions
        checks.append(
            {
                "check": "min_interruptions",
                "pass": ok,
                "expected": verify.min_interruptions,
                "actual": len(interruptions),
            }
        )
    if verify.max_interruptions is not None:
        ok = len(interruptions) <= verify.max_interruptions
        checks.append(
            {
                "check": "max_interruptions",
                "pass": ok,
                "expected": verify.max_interruptions,
                "actual": len(interruptions),
            }
        )

    plugin_results: list[dict[str, object]] = []
    if verify.plugins:
        from ..plugins.api import VerifyContext
        from ..plugins.loader import ensure_plugins_loaded
        from ..plugins.registry import get_verify

        if project_root is not None:
            ensure_plugins_loaded(
                project_root,
                list(scenario.plugin_modules) if scenario is not None else None,
            )
        for plugin_name in verify.plugins:
            fn = get_verify(plugin_name)
            if fn is None:
                checks.append(
                    {
                        "check": f"plugin:{plugin_name}",
                        "pass": False,
                        "reason": f"verify plugin {plugin_name!r} is not registered",
                    }
                )
                continue
            if scenario is None or project_root is None:
                checks.append(
                    {
                        "check": f"plugin:{plugin_name}",
                        "pass": False,
                        "reason": "plugin verify requires scenario and project_root",
                    }
                )
                continue
            opts = verify.plugin_options.get(plugin_name, {})
            if not isinstance(opts, dict):
                opts = {}
            ctx = VerifyContext(
                events=events,
                steps=steps,
                verify=verify,
                scenario=scenario,
                project_root=Path(project_root),
                plugin_name=plugin_name,
                options=dict(opts),
            )
            try:
                raw = fn(ctx)
            except Exception as e:
                checks.append(
                    {
                        "check": f"plugin:{plugin_name}",
                        "pass": False,
                        "reason": f"{type(e).__name__}: {e}",
                    }
                )
                continue
            passed = bool(raw.get("pass"))
            plugin_checks = raw.get("checks")
            if isinstance(plugin_checks, list):
                for item in plugin_checks:
                    if isinstance(item, dict):
                        checks.append({**item, "plugin": plugin_name})
            checks.append(
                {
                    "check": f"plugin:{plugin_name}",
                    "pass": passed,
                    "plugin": plugin_name,
                    "detail": raw.get("detail"),
                }
            )
            plugin_results.append({"plugin": plugin_name, "pass": passed, "result": raw})

    return {
        "script_steps": len(steps),
        "cues_fired": len([e for e in cues if e.get("kind") == "sim.script.cue"]),
        "waits_fired": len([e for e in cues if e.get("kind") == "sim.script.wait"]),
        "hang_ups_fired": len([e for e in cues if e.get("kind") == "sim.script.hang_up"]),
        "agent_finals_after_first_cue": agent_after_cue,
        "user_finals_after_first_cue": user_after_cue,
        "agent_finals_after_silence": agent_after_silence,
        "agent_finals_after_barge_in": agent_after_barge,
        "interruptions": len(interruptions),
        "checks": checks,
        "plugin_results": plugin_results,
        "pass": all(bool(c.get("pass")) for c in checks) if checks else False,
    }

File: src/livekit_agent_simulator_rust/crates/lks-core/src/script/verify.rs
//! Log-based script verify — byte-parity port of `script/verify.py` (core,
//! non-plugin path). Plugin checks land in P8 (pyo3 loader).

use serde_json::{json, Map, Value as Json};

use super::{counts_for_recovery_barge, ScriptStep, ScriptVerifySpec};

fn ev_kind(e: &Map<String, Json>) -> &str {
    e.get("kind").and_then(|v| v.as_str()).unwrap_or("")
}

/// The spec dict of an event (shared empty map when absent).
fn ev_spec(e: &Map<String, Json>) -> &Map<String, Json> {
    static EMPTY: std::sync::OnceLock<Map<String, Json>> = std::sync::OnceLock::new();
    e.get("spec")
        .and_then(|v| v.as_object())
        .unwrap_or_else(|| EMPTY.get_or_init(Map::new))
}

fn ev_mono(e: &Map<String, Json>) -> i64 {
    e.get("ts_mono_ms").and_then(|v| v.as_i64()).unwrap_or(0)
}

/// Evaluate a script log against steps + verify spec.
/// Returns the same dict shape as `evaluate_script_log`.
pub fn evaluate_script_log(
    events: &[Map<String, Json>],
    steps: &[ScriptStep],
    verify: Option<&ScriptVerifySpec>,
) -> Map<String, Json> {
    let cues: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| {
            let k = ev_kind(e);
            k == "sim.script.cue" || k == "sim.script.wait" || k == "sim.script.hang_up"
        })
        .collect();
    let agent_finals: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "transcript.agent.final")
        .collect();
    let user_finals: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "transcript.user.final")
        .collect();
    let interruptions: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "interruption")
        .collect();

    let mut checks: Vec<Json> = Vec::new();

    // Per-step checks.
    for step in steps {
        let matching: Vec<&&Map<String, Json>> = cues
            .iter()
            .filter(|c| {
                ev_spec(c).get("step_id").and_then(|v| v.as_str()) == Some(step.id.as_str())
            })
            .collect();
        if matching.is_empty() {
            checks.push(json!({
                "step_id": step.id,
                "pass": false,
                "reason": "script step not fired",
            }));
            continue;
        }
        let cue = matching[0];
        let spec = ev_spec(cue);
        if let Some(err) = spec.get("error") {
            checks.push(json!({
                "step_id": step.id,
                "pass": false,
                "reason": format!("cue fired with error: {err}"),
            }));
            continue;
        }
        let during = spec
            .get("during_agent_speech")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        if step.trigger == "agent_speaking" && step.action == "speak" && !during {
            checks.push(json!({
                "step_id": step.id,
                "pass": false,
                "reason": "cue fired but agent was not active speaker",
            }));
            continue;
        }
        checks.push(json!({
            "step_id": step.id,
            "pass": true,
            "during_agent_speech": during,
            "trigger": step.trigger,
            "action": step.action,
        }));
    }

    let cue_ms = cues.first().map(|c| ev_mono(c));
    let silence_cues: Vec<&&Map<String, Json>> = cues
        .iter()
        .filter(|e| {
            let s = ev_spec(e);
            s.get("trigger").and_then(|v| v.as_str()) == Some("silence")
                || s.get("action").and_then(|v| v.as_str()) == Some("wait")
        })
        .collect();
    let silence_ms = silence_cues.first().map(|c| ev_mono(c));

    // Barge cues: counts_for_recovery_barge OR legacy short-during-agent heuristic.
    let mut barge_cues: Vec<&&Map<String, Json>> = Vec::new();
    for e in &cues {
        if ev_kind(e) != "sim.script.cue" {
            continue;
        }
        let spec = ev_spec(e);
        let cls = spec
            .get("class")
            .or_else(|| spec.get("interrupt_class"))
            .map(|v| v.as_str().unwrap_or(&v.to_string()).to_string());
        if counts_for_recovery_barge(
            spec.get("barge_in")
                .and_then(|v| v.as_bool())
                .unwrap_or(false),
            cls.as_deref(),
        ) {
            barge_cues.push(e);
            continue;
        }
        // Legacy heuristic: short during-agent cue without class → correction.
        if cls.is_none()
            && spec
                .get("during_agent_speech")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
            && spec.get("trigger").and_then(|v| v.as_str()) == Some("agent_speaking")
            && spec
                .get("waited_ms")
                .and_then(|v| v.as_i64())
                .unwrap_or(9999)
                < 800
        {
            barge_cues.push(e);
        }
    }
    let barge_ms = barge_cues.first().map(|c| ev_mono(c));

    let count_ge = |finals: &[&Map<String, Json>], ms: Option<i64>| -> i64 {
        match ms {
            Some(ms) => finals.iter().filter(|e| ev_mono(e) >= ms).count() as i64,
            None => 0,
        }
    };
    let count_gt = |finals: &[&Map<String, Json>], ms: Option<i64>| -> i64 {
        match ms {
            Some(ms) => finals.iter().filter(|e| ev_mono(e) > ms).count() as i64,
            None => 0,
        }
    };
    let agent_after_cue = count_ge(&agent_finals, cue_ms);
    let user_after_cue = count_ge(&user_finals, cue_ms);
    let agent_after_silence = count_ge(&agent_finals, silence_ms);
    let agent_after_barge = count_gt(&agent_finals, barge_ms);

    let verify = verify.cloned().unwrap_or(ScriptVerifySpec {
        require_during_agent_speech: true,
        min_agent_finals_after_first_cue: 0,
        min_user_finals_after_first_cue: 0,
        min_interruptions: None,
        max_interruptions: None,
        min_agent_finals_after_silence: 0,
        min_agent_finals_after_barge_in: 0,
        plugins: Vec::new(),
        plugin_options: Map::new(),
    });

    if verify.min_agent_finals_after_first_cue > 0 {
        checks.push(json!({
            "check": "min_agent_finals_after_first_cue",
            "pass": agent_after_cue >= verify.min_agent_finals_after_first_cue,
            "expected": verify.min_agent_finals_after_first_cue,
            "actual": agent_after_cue,
        }));
    }
    if verify.min_user_finals_after_first_cue > 0 {
        checks.push(json!({
            "check": "min_user_finals_after_first_cue",
            "pass": user_after_cue >= verify.min_user_finals_after_first_cue,
            "expected": verify.min_user_finals_after_first_cue,
            "actual": user_after_cue,
        }));
    }
    if verify.min_agent_finals_after_silence > 0 {
        checks.push(json!({
            "check": "min_agent_finals_after_silence",
            "pass": agent_after_silence >= verify.min_agent_finals_after_silence,
            "expected": verify.min_agent_finals_after_silence,
            "actual": agent_after_silence,
        }));
    }
    if verify.min_agent_finals_after_barge_in > 0 {
        checks.push(json!({
            "check": "min_agent_finals_after_barge_in",
            "pass": agent_after_barge >= verify.min_agent_finals_after_barge_in,
            "expected": verify.min_agent_finals_after_barge_in,
            "actual": agent_after_barge,
        }));
    }
    if let Some(min_i) = verify.min_interruptions {
        checks.push(json!({
            "check": "min_interruptions",
            "pass": (interruptions.len() as i64) >= min_i,
            "expected": min_i,
            "actual": interruptions.len(),
        }));
    }
    if let Some(max_i) = verify.max_interruptions {
        checks.push(json!({
            "check": "max_interruptions",
            "pass": (interruptions.len() as i64) <= max_i,
            "expected": max_i,
            "actual": interruptions.len(),
        }));
    }

    let cues_fired = cues
        .iter()
        .filter(|e| ev_kind(e) == "sim.script.cue")
        .count();
    let waits_fired = cues
        .iter()
        .filter(|e| ev_kind(e) == "sim.script.wait")
        .count();
    let hang_ups_fired = cues
        .iter()
        .filter(|e| ev_kind(e) == "sim.script.hang_up")
        .count();
    let all_pass = if checks.is_empty() {
        false
    } else {
        checks
            .iter()
            .all(|c| c.get("pass").and_then(|v| v.as_bool()).unwrap_or(false))
    };

    let mut out = Map::new();
    out.insert("script_steps".into(), json!(steps.len()));
    out.insert("cues_fired".into(), json!(cues_fired));
    out.insert("waits_fired".into(), json!(waits_fired));
    out.insert("hang_ups_fired".into(), json!(hang_ups_fired));
    out.insert(
        "agent_finals_after_first_cue".into(),
        json!(agent_after_cue),
    );
    out.insert("user_finals_after_first_cue".into(), json!(user_after_cue));
    out.insert(
        "agent_finals_after_silence".into(),
        json!(agent_after_silence),
    );
    out.insert(
        "agent_finals_after_barge_in".into(),
        json!(agent_after_barge),
    );
    out.insert("interruptions".into(), json!(interruptions.len()));
    out.insert("checks".into(), Json::Array(checks));
    out.insert("plugin_results".into(), Json::Array(Vec::new()));
    out.insert("pass".into(), json!(all_pass));
    out
}
